eZ/Publish/Core/Repository/PermissionsCriterionHandler.php

```diff
@@ -118,6 +118,7 @@ public function getPermissionsCriterion($module = 'content', $function = 'read')
             foreach ($permissionSet['policies'] as $policy) {
                 $limitations = $policy->getLimitations();
                 if ($limitations === '*' || empty($limitations)) {
+                    $policyOrCriteria[] = new Criterion\MatchAll();
```

Comment:
I think we should be able to optimize this, seems wasteful to handle it this way. Maybe we should eventually introduce some kind of optimizer service to do this on role assignments (FWIW I would rather implement analyzer command then add more logic to the core), but from the POV of search, it's quite obvious what should be done.

Then, if there are no limitations, I think it should be OK to do this:

```php
$policyOrCriteria = [];
break;
```

This depends on the assumption that **there IS a role limitation**, otherwise `hasAccess()` would have returned `true` and this would've not been reached.